File: internal/cmd/new.go
package cmd

import (
	"fmt"
	"path/filepath"
	"strings"

	"github.com/0xkhdr/specd/internal/cli"
	"github.com/0xkhdr/specd/internal/core"
)

func titleCase(slug string) string {
	parts := strings.Split(slug, "-")
	for i, p := range parts {
		if len(p) > 0 {
			parts[i] = strings.ToUpper(p[:1]) + p[1:]
		}
	}
	return strings.Join(parts, " ")
}

// RunNew implements `specd new`: it validates the slug, refuses to overwrite
// an existing spec, materializes the six spec artifacts from templates
// (seeding requirements.md with --from, if given), writes the initial
// state.json (optionally starting in orchestrated mode with --orchestrated),
// and prints the next step.
func RunNew(args cli.Args) int {
	root, slug, code, ok := requireRootAndSlug(args, "usage: specd new <slug> [--title \"...\"]")
	if !ok {
		return code
	}
	if err := core.ValidateSlug(slug); err != nil {
		return specdExit(err)
	}
	if core.SpecExists(root, slug) {
		core.Error(fmt.Sprintf("spec '%s' already exists", slug))
		return core.ExitGate
	}
	title := args.Str("title")
	if title == "" {
		title = titleCase(slug)
	}
	date := core.Clock().UTC().Format("2006-01-02")
	vars := map[string]string{"TITLE": title, "SLUG": slug, "DATE": date}
	prompt := strings.TrimSpace(args.Str("from"))

	dir := core.SpecDir(root, slug)
	for _, name := range core.Artifacts {
		tmplPath := "specStubs/" + name
		tmpl, err := core.ReadTemplate(tmplPath)
		if err != nil {
			core.Error(fmt.Sprintf("missing template %s: %v", tmplPath, err))
			return core.ExitGate
		}
		content := core.ApplyVars(tmpl, vars)
		if name == "requirements.md" {
			content = core.InjectPrompt(content, prompt)
		}
		if err := core.AtomicWrite(filepath.Join(dir, name), content); err != nil {
			core.Error(fmt.Sprintf("write %s: %v", name, err))
			return core.ExitGate
		}
	}
	state := core.InitialState(slug, title)
	state.Prompt = prompt
	if args.Bool("orchestrated") {
		if !core.ProjectOrchestrationEnabled(root) {
			core.Error("--orchestrated: project has no orchestration capability. Enable it with `specd init --orchestration session` (or manual|planning), then retry.")
			return core.ExitGate
		}
		state.ExecutionMode = core.ModeOrchestrated
		state.ModeOrigin = core.OriginUser
	}
	if err := core.SaveState(root, slug, &state); err != nil {
		return specdExit(err)
	}
	fmt.Printf("specd new: created spec '%s' (%s)\n", slug, title)
	fmt.Printf("  .specd/specs/%s/ — six artifacts + state.json (status: requirements)\n", slug)
	fmt.Printf("Next: write requirements.md (EARS), then `specd check %s`.\n", slug)
	return core.ExitOK
}

File: src/commands/new.ts
// `specd new <slug> [--title "..."]` — create a spec folder with six artifacts + state.json (§5.2).
import { join } from "node:path";
import { atomicWrite } from "../core/io.js";
import { EXIT, usageError, gateError, SpecdError } from "../core/exit.js";
import { requireSpecdRoot, specDir } from "../core/paths.js";
import { ARTIFACTS, specExists } from "../core/specFiles.js";
import { initialState, saveState } from "../core/state.js";
import { applyVars, readTemplate } from "../core/templates.js";
import type { Args } from "../cli.js";

const SLUG_RE = /^[a-z0-9][a-z0-9-]*$/;

const titleCase = (slug: string) =>
  slug.split("-").map((w) => w.charAt(0).toUpperCase() + w.slice(1)).join(" ");

export function run(args: Args): number {
  const root = requireSpecdRoot();
  const slug = args.pos[0];
  if (!slug) throw usageError("usage: specd new <slug> [--title \"...\"]");
  if (!SLUG_RE.test(slug)) throw usageError(`invalid slug '${slug}' (must match ^[a-z0-9][a-z0-9-]*$)`);
  if (specExists(root, slug)) throw new SpecdError(EXIT.GATE, `spec '${slug}' already exists`);

  const title = typeof args.flags.title === "string" ? args.flags.title : titleCase(slug);
  const date = new Date().toISOString().slice(0, 10);
  const vars = { TITLE: title, SLUG: slug, DATE: date };

  const dir = specDir(root, slug);
  for (const name of ARTIFACTS) {
    const tmpl = readTemplate(`specStubs/${name}`);
    atomicWrite(join(dir, name), applyVars(tmpl, vars));
  }
  saveState(root, slug, initialState(slug, title));

  console.log(`specd new: created spec '${slug}' (${title})`);
  console.log(`  .specd/specs/${slug}/ — six artifacts + state.json (status: requirements)`);
  console.log("Next: write requirements.md (EARS), then `specd check " + slug + "`.");
  return 0;
}
